### ISSUE TITLE
Invalid argument supplied for foreach() /usr/share/dokuwiki/lib/exe/js.php

### ISSUE THREAD

**Issue by niffumau** [splitbrain/dokuwiki#1726]
Invalid argument supplied for foreach() /usr/share/dokuwiki/lib/exe/js.php.

Not sure if this is something to do with my <B>FrankinDebian</b> install or not but when it evaluates "foreach($config_cascade['userscript']['default'] as $userscript)" if "$config_cascade['userscript']['default']" only has one thing in the list, it seems to be treated no longer as an array.

I also noticed this happens some where else but i can't find my notes where that is.

I modified:
  // add possible plugin scripts and userscript
    $files   = array_merge($files,js_pluginscripts());
    if(!empty($config_cascade['userscript']['default'])) {
        foreach($config_cascade['userscript']['default'] as $userscript) {
            $files[] = $userscript;
        }
    }

to:
// add possible plugin scripts and userscript
    $files   = array_merge($files,js_pluginscripts());
    if(!empty($config_cascade['userscript']['default'])) {
        if (is_array($config_cascade['userscript']['default']))
                foreach($bullshit as $userscript) {
                    $files[] = $userscript;
                } else $files[] = $config_cascade['userscript']['default'];
    }

**Comment by Klap-in** [comment]
Looks like the same issue as https://github.com/splitbrain/dokuwiki/issues/1664
So probably it useful to check if the Debian version requires attention for this issue. At https://www.dokuwiki.org/install:debian there is link to the issue tracker of the maintainer of this package.

**Comment by niffumau** [comment]
yes, that is exactly the error. Errors i think are usually suppressed so it just wouldn't load the js file and you wouldn't know any different.  Its evaluating something that is not an array as an array.

**Comment by Klap-in** [comment]
Do you still run a version of DokuWiki as delivered via Debian repository? or did you manually update it with a package downloaded from dokuwiki.org?

**Comment by niffumau** [comment]
Oh right, i see, yes,  /usr/share/dokuwiki/inc/preload.php is a debian thing, I didn't realize..
In there they declare 
 'userscript' => array(
        'default' => DOKU_CONF.'userscript.js'
    ),
which should be then
 'userscript' => array(
        'default' => array(DOKU_CONF.'userscript.js')
    ),
although some checking if it was an array would be handy.

I'll see if i can raise a bug over there, thanks heaps.

**Comment by Klap-in** [comment]
This is broken since: https://github.com/splitbrain/dokuwiki/pull/954

My guess is that single value entries from `$config_cascade` overwrites the new array shaped entries from `$config_cascade` (https://github.com/splitbrain/dokuwiki/blob/master/inc/config_cascade.php)
So entries are just single valued entities instead of arrays. 

A solution could be to introduce somewhere casting to array. 

Question: 
has casting still disadvantages? That people do not update to new format? or...

**Comment by splitbrain** [comment]
I think this should be fixed in Debian.

**Comment by Klap-in** [comment]
The example/default files should be fixed, like already done for the farms indeed.
I now see that a `preload.php` is provided in the package. 

@ortolot do you have chance to have a look on this?
Note that some other entries are modified in the `config_cascade.php` recently too. So maybe these are useful to change in preload.php as well.

**Comment by fumiyas** [comment]
I've posted a bug report to Debian BTS:

* #866245 - dokuwiki: Invalid argument supplied for foreach() /usr/share/dokuwiki/lib/exe/js.php - Debian Bug report logs
  * https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=866245
